### DISCUSSION TITLE
Livewire 3

### DISCUSSION THREAD

**Discussion by luanfreitasdev** [Power-Components/livewire-powergrid#1072] [Announcements]
WIP

**Comment by luanfreitasdev** [comment]
Researching performance issue: https://github.com/livewire/livewire/discussions/6052

**Reply by luanfreitasdev** [reply]
Done 😃 (_No real issues were encountered with the latest livewire modifications._)

https://github.com/Power-Components/livewire-powergrid/assets/33601626/846a1b05-eaa9-4a47-b9d7-621d7931396c

**Comment by luanfreitasdev** [comment]
### Some breakpoints for the new version with livewire 3 and improvements:

* Ended support for version 2.
* The method "_actions_" and "_actionRules_" will have row scope (Model, array).

```php
// v4
public function actions(): array
{
}

public function actionsRules(): array
{
}

// v5
public function actions(Dish $dish): array
{
}

public function actionsRules(Dish $dish): array
{
}
```

* Reading into the model along with parameters will no longer work, because $row is now in the scope of the function

```php
## old
Button::add('create-dish')  
    ->caption('Create a dish')
    ->class('bg-indigo-500 text-white')
    ->emit('postAdded', ['key' => 'id']), // deprecated

## new
Button::add('create-dish')  
    ->caption('Create a dish')
    ->class('bg-indigo-500 text-white')
    ->dispatch('postAdded', ['key' =>$row->id]), // <-----------------

```

* Creating a "::action" column is required

```php
             Column::make('Created At', 'created_at_formatted'),
             Column::action('Action'), // <-----------------
        ];
    }
```

* All actions (except id, class, caption) will only be performed via macro, allowing many customizations.

### Macros added
```php
/**
 * @method static dispatch(string $event, array $params)
 * @method static dispatchTo(string $component, string $event, array $params)
 * @method static dispatchSelf(string $event, array $params)
 * @method static openModal(string $component, array $params)
 * @method static parent(string $method, array $params)
 * @method static call(string $method, array $params)
 * @method static toggleDetail()
 * @method static tooltip(string $title)
 * @method static route(string $route, array $params)
 * @method static method(string $method)
 * @method static target(string $target) _blank, _self, _top, _parent, null
 * @method static render(\Closure $closure)
 * @method static bladeComponent(string $component, array $params)
 *
 */
```

**Reply by luanfreitasdev** [reply]
Improvement suggestions are also welcome.

**Reply by hitaloramon** [reply]
@luanfreitasdev 

What I would like to suggest are visual improvements.

1. The mass action button could only be displayed when it had a row selected and if it had more than 1 mass action button, it could be displayed within a dropdown, because for those who have many actions it would be better visually.

2. Filters could also have a way to display within a dropdown

Exemple: 

<img width="988" alt="bulk-actions" src="https://github.com/Power-Components/livewire-powergrid/assets/12436779/917576b2-6074-40eb-b21f-f91d7c42cb03">

**Reply by luanfreitasdev** [reply]
Do you think of any different code syntax?

**Reply by hitaloramon** [reply]
I think the class syntax is good, it's very easy to create. I think we just need to improve the visualization. Check how many bulk actions there are and if it's greater than 2, change the display to suspended; if it is smaller, it only displays the button

**Reply by CrazyBoy49z** [reply]
@luanfreitasdev Where can I find the repository for Powergrid v5? I urgently need support for Livewire 3. The only obstacle is Powergrid.

**Reply by luanfreitasdev** [reply]
Hello @CrazyBoy49z ! I'm also looking forward to releasing a stable version to work with Livewire 3. But even Livewire 3 isn't stable, so Powergrid can't be an obstacle as Livewire 3 isn't ready for production yet.

But I understand your need, and I'm working to make the 5.x-dev version available, but the work is still in progress.

**Reply by CrazyBoy49z** [reply]
I am eagerly awaiting the dev version

**Reply by luanfreitasdev** [reply]
It's not in beta yet, but we can test it, feel free to report any issues (in this thread), open a pull request for a fix, help with the documentation.

![image](https://github.com/Power-Components/livewire-powergrid/assets/33601626/4ffbfcfa-77e8-47bd-a0cf-17ac99a015d0)

https://github.com/Power-Components/livewire-powergrid/tree/5.x

**Comment by luanfreitasdev** [comment]
I made a nice improvement here that will be available in this version

https://github.com/Power-Components/livewire-powergrid/commit/3b4b79c8c9a5fa602c4e50c40366c79507eb9677

* Performance improvement (does not use get());
* Custom format

```php
 public function summarizeFormat(): array
 {
      return [
          'price' => function ($value) {
              return (new \NumberFormatter('en_US', \NumberFormatter::CURRENCY))
                  ->formatCurrency($value, 'USD');
           }
      ];
}

public function columns()
{
      return [
            Column::make('Price', 'price')
                ->withSum('Sum Price', true, false)
                ->withCount('Count Price', true, false)
                ->withAvg('Avg Price', true, false)
                ->sortable(),
      ];
}
```

![image](https://github.com/Power-Components/livewire-powergrid/assets/33601626/caae3122-00b5-4d1a-9266-283c2972af2f)
![image](https://github.com/Power-Components/livewire-powergrid/assets/33601626/8638247a-3f11-489c-9aa7-782765068603)

**Comment by luanfreitasdev** [comment]
**Added filterRelation** - https://github.com/Power-Components/livewire-powergrid/commit/2778bd113305bf4099c6cf48e37c17c4d23d2194

```php
public function filters(): array
{
    return [
        Filter::inputText('category_name')
           ->filterRelation('category', 'name')
     ];
}
```

**Comment by luanfreitasdev** [comment]
Hey everyone! I released a pre-release and by making final adjustments we can launch the official soon.

https://github.com/Power-Components/livewire-powergrid/releases/tag/v5.0.0-beta

How to launch faster?
You can help by testing, submitting bug reports, and sending pull requests to the 5.x branch.

Thanks
